src/back/hlsl/writer.rs

```diff
@@ -1441,6 +1566,180 @@ impl<'a, W: Write> Writer<'a, W> {
+                                    _ => match (dim, class) {
+                                        (_, crate::ImageClass::Sampled { multi: true, .. }) => {}
+                                        (IDim::D2 | IDim::D3 | IDim::Cube, _) => {
+                                            write!(self.out, "0, ")?;
+                                        }
+                                        _ => {}
+                                    },
+                                }
+                                for i in 0..number_of_params {
+                                    if i == (number_of_params - 1) {
+                                        // write last parameter without comma and space for last parameter
+                                        write!(
+                                            self.out,
+                                            "{}.{}",
+                                            RETURN_VARIABLE_NAME,
+                                            back::COMPONENTS[i]
+                                        )?;
+                                    } else {
+                                        write!(
+                                            self.out,
+                                            "{}.{}, ",
+                                            RETURN_VARIABLE_NAME,
+                                            back::COMPONENTS[i]
+                                        )?;
+                                    }
+                                }
+                                writeln!(self.out, ");")?;
+
+                                // Write return value
+                                writeln!(
+                                    self.out,
+                                    "{}return {}.{};",
+                                    back::INDENT,
+                                    RETURN_VARIABLE_NAME,
+                                    ret_swizzle
+                                )?;
+
+                                // End of function body
+                                writeln!(self.out, "}}")?;
+                                // Write extra new line
+                                writeln!(self.out)?;
+
+                                self.wrapped_image_queries.insert(wrapped_image_query);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn write_wrapped_image_query_function_name(
+        &mut self,
+        query: WrappedImageQuery,
+    ) -> BackendResult {
+        write!(
+            self.out,
+            "NagaGetDimensions{}{}{}{}",
+            image_dimension_str(query.dim),
+            match query.class {
+                crate::ImageClass::Sampled { multi: true, .. } => "MS",
+                crate::ImageClass::Depth => "Depth",
+                _ => "",
+            },
+            if query.arrayed { "Array" } else { "" },
+            match query.query {
```

Comment:
Let's not put any logic inside the `write!` macro invocation. For any tools the macros are harder to work with, so we should take advantage of plain code where possible.